packages/util/src/errors.ts

```diff
@@ -54,115 +54,95 @@
-          }
-        });
-      }
-    }
-  }
+  // Additional custom error data that was used in the template.
+  readonly data: ErrorData;
 }
 
-// Back-door inheritance
-FirebaseError.prototype = Object.create(Error.prototype) as FirebaseError;
-FirebaseError.prototype.constructor = FirebaseError;
-(FirebaseError.prototype as any).name = ERROR_NAME;
-
-export class ErrorFactory<T extends string> {
-  // Matches {$name}, by default.
-  public pattern = /\{\$([^}]+)}/g;
+// Based on code from:
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#Custom_Error_Types
+export class FirebaseError extends Error {
+  readonly name = ERROR_NAME;
 
   constructor(
-    private service: string,
-    private serviceName: string,
-    private errors: ErrorList<T>
+    readonly code: string,
+    message: string,
+    readonly data: ErrorData = {}
   ) {
-    // empty
-  }
+    super(message);
+
+    // Fix For ES5
+    // https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
+    Object.setPrototypeOf(this, FirebaseError.prototype);
 
-  create(code: T, data?: { [prop: string]: StringLike }): FirebaseError {
-    if (data === undefined) {
-      data = {};
+    // Maintains proper stack trace for where our error was thrown.
+    // Only available on V8.
+    if (Error.captureStackTrace) {
+      Error.captureStackTrace(this, ErrorFactory.prototype.create);
     }
+  }
+}
 
-    let template = this.errors[code as string];
+export class ErrorFactory<ErrorCode extends string> {
+  constructor(
+    private readonly service: string,
+    private readonly serviceName: string,
+    private readonly errors: ErrorList<ErrorCode>
+  ) {}
 
-    let fullCode = this.service + '/' + code;
-    let message: string;
+  create(code: ErrorCode, data: ErrorData = {}): FirebaseError {
+    const fullCode = `${this.service}/${code}`;
+    const template = this.errors[code];
 
-    if (template === undefined) {
-      message = 'Error';
-    } else {
-      message = template.replace(this.pattern, (match, key) => {
-        let value = data![key];
-        return value !== undefined ? value.toString() : '<' + key + '?>';
-      });
-    }
+    const message = template ? replaceTemplate(template, data) : 'Error';
 
     // Service: Error message (service/code).
-    message = this.serviceName + ': ' + message + ' (' + fullCode + ').';
-    let err = new FirebaseError(fullCode, message);
-
-    // Populate the Error object with message parts for programmatic
-    // accesses (e.g., e.file).
-    for (let prop in data) {
-      if (!data.hasOwnProperty(prop) || prop.slice(-1) === '_') {
-        continue;
+    const fullMessage = `${this.serviceName}: ${message} (${fullCode}).`;
+
+    // Keys with an underscore at the end of their name are not included in
+    // error.data for some reason.
+    const filteredData: ErrorData = {};
```

Comment:
Oh man, this was my favorite part :disappointed:

I thought it was okay in this case, since it's a private API anyway. I guess our users would be able to access these fields if they catch a `FirebaseError`, but I don't think we have that documented anywhere? Unless an SDK is actually saying "This method will throw an Error with `x` and `y` fields set", I don't think this would count as an API change.

I'm gonna revert this part for now, but what about at least doing a `console.warn` if a field in `data` overwrites `code`, `message`, `name`, or `stack`? I'd say it's still bad since different browsers might have different fields in their `Error` implementation (apparently only `name` and `message` are part of the spec, the rest are made up, including `stack`), but it's still better than the current version (which caused me to spend some time before I realized I was overwriting `code` and `message` in my template once).

Reply:
> we can have a similar proposal for the Web SDK

Yes please, anything that makes our packages more consistent is :+1: